refactor observer patterns and unify logging information

tarun-mitruka opened the pull request:
This PR heavily refactors the interface of the observer patterns. By the addition of `struct ControlLoggerInformation` and `struct NonLinearSolverLoggingInformation`, the function `updateImpl` of the class `IObserver` can be simplified by just passing these `struct` and thus removes the overloading of this function. This simplifies the whole interface of the class `IObservable` as well. Moreover, the `switch` statements in `controllogger.cpp` and `nonlinearsolverlogger.cpp` can be combined in one function which increases code readability and extensibility. 
The information of the load factor `lambda` is now also moved from a solver logger to a control logger as it is expected to be. This leads to small fixes in the class `LoadControl` which now then shows the same output as by the class `PathFollowing` as confirmed by an additional test added in `testpathfollowing.cpp`.

rath3t commented on docs/website/01_framework/observer.md:
```diff
@@ -26,38 +26,45 @@ enum class NonLinearSolverMessages {
+To become an observer, the class must inherit from ``IObserver<MessageType, InfoType>``.
+Here, ``MessageType``
+is the `enum` of messages to use (see above)
+and ``InfoType`` can either be ``NonLinearSolverLoggingInformation`` or ``ControlLoggerInformation``.
```
Ifno type can be anything, since it is a template. It has to adhere to some interface, e.g. maybe a concept but now this is just some tempalte isn't it?

rath3t commented on docs/website/01_framework/observer.md:
````diff
@@ -85,13 +92,17 @@ unSubscribe(...) // (4)!
 
 ```cpp
-this->notify(MessageType::Message, data);
+this->notify(MessageType message, const InfoType& info);
````
This code is at callside, therefore there should be now type information. If you want to indicate it you can make more lines wher you create the objects. 

```cpp

InfoType info;
this->notify(message,  info);
```
or similar

rath3t commented on ikarus/controlroutines/controlinfos.hh:
```diff
@@ -19,10 +19,25 @@ namespace Ikarus {
+};
+
+/**
+ * \struct ControlLoggerInformation
```
These things are not `ControlLoggerInformation` they are actually information of the control routine.

Maybe we should refactor the naming. Anyway the observers now have a `message` which is a enum and independent of the message, the observers always receive this struct. 

Therefore, maybe, if we want to keep the message notion, we can call this `ControlLoggerInformation` --> `ControlState` ?

rath3t commented on ikarus/controlroutines/loadcontrol.hh:
```diff
@@ -51,6 +51,8 @@ public:
           nonLinearSolver_->nonLinearOperator().lastParameter() = 0.0;
           nonLinearSolver_->nonLinearOperator().lastParameter() += 0.0;
         }, "The last parameter (load factor) must be assignable and incrementable with a double!");
+    if (loadSteps_ <= 0)
```
maybe make loadstep `unsigned int` and only check for zero equality?

rath3t commented on ikarus/solver/nonlinearsolver/newtonraphson.hh:
```diff
@@ -115,21 +116,27 @@ public:
-      this->notify(NonLinearSolverMessages::SOLUTION_CHANGED);
+      logInfo.correctionNorm = static_cast<double>(dNorm);
+      this->notify(NonLinearSolverMessages::CORRECTIONNORM_UPDATED, logInfo);
+      this->notify(NonLinearSolverMessages::SOLUTION_CHANGED, logInfo);
```
I can already see the improvement. This is so much cleaner now!

tarun-mitruka replied:
Thanks

rath3t commented on ikarus/solver/nonlinearsolver/solverinfos.hh:
```diff
@@ -23,9 +23,23 @@ struct NonLinearSolverInformation
+ * \details This structure holds information about the value of the load factor, the residual norm,
+ * the correction norm and the number of iterations performed.
+ */
+struct NonLinearSolverLoggingInformation
```
Same discussion as above

rath3t replied:
But maybe both things should be merged anyway. `NonLinearSolverInformation` and the logging information into one `State` or `info` struct.

rath3t reviewed (commented):
Thansk for the PR. We can also diccuss the changes. Then I will do the full review.

You dnot have to refactor yet.

tarun-mitruka commented on ikarus/controlroutines/controlstate.hh:
```diff
@@ -0,0 +1,38 @@
+  std::vector<NonLinearSolverState>
+      solverState{};   ///< Vector containing information about the state of the nonlinear solver.
+  std::string name{};  ///< Information about the name of the control method.
+  Eigen::VectorXd sol; ///< The solution vector, for example, the displacement vector.
```
Here, instead of having the `NonLinearOperator` as we discussed, I thought it gives better flexibility and readability perhaps to store its first and last parameter separately. Do you think it is better to have the non-linear operator instead and remove `lambda` and `sol`?

tarun-mitruka commented on ikarus/controlroutines/loadcontrol.hh:
```diff
@@ -51,21 +51,41 @@ public:
+   * \param nonOp The non-linear operator.
+   * \param solverState State of the nonlinear solver.
+   */
+  void updateAndNotifyControlState(ControlState& controlState, typename NLS::NonLinearOperator& nonOp,
```
This function just wraps certain repeated statements for better readability in `loadcontrol.inl`. The same is now implemented in `pathfollowing.inl`. I did not want to pass the control routine itself and have this function as a free function in some helper file, but perhaps the necessity of this as a free function in a separate file can be discussed.

tarun-mitruka replied:
We keep this as such for now

tarun-mitruka commented on ikarus/controlroutines/pathfollowing.inl:
```diff
@@ -23,51 +23,54 @@ namespace Ikarus {
-
-  /// Calculate predictor for a particular step
-  for (int ls = 1; ls < steps_; ++ls) {
+  /// Dummy execution of the code inorder to save information of the undeformed (or initial)
```
These parts of the code are modified only because this PR also unifies the output of `LoadControl` and `PathFollowing`, which affects the interface the observer patterns and hence it is not split into two PRs

tarun-mitruka commented on ikarus/controlroutines/pathfollowingfunctions.hh:
```diff
@@ -74,13 +74,11 @@ struct ArcLength
   void operator()(SubsidiaryArgs& args) const {
-    if (psi) {
-      const auto root = sqrt(args.DD.squaredNorm() + psi.value() * psi.value() * args.Dlambda * args.Dlambda);
+    if (psi != std::numeric_limits<double>::infinity()) {
```
I removed `psi` as an optional variable. This is because `initialPrediction` is not called for the dummy execution of the code to save the state of the initial configuration, which then throws an error when the `else` branch is called. But, still to ensure a reasonable update of `psi`, I defaulted its value to infinity, such that numerical problems arise if the it is not updated later at the first load step.

tarun-mitruka replied:
I changed it back to the original version as now the interface of control routines doesn't solve the system of equations when executing the dummy steps to save the information of the initial configuration.

tarun-mitruka commented on ikarus/utils/observer/controllogger.cpp:
```diff
@@ -9,51 +9,42 @@
       break;
-    default:
+    case ControlMessages::STEP_STARTED:
+      if (state.initialConfig) {
```
A boolean to flag if the configuration is an initial configuration `initialConfig` is introduced such that here the output is differentiated and is reasonable, else the numbers displayed on the screen can be misleading

tarun-mitruka commented on ikarus/utils/observer/observer.hh:
```diff
@@ -138,169 +77,84 @@ public:
-template <typename MT>
-void IObservable<MT>::subscribe(MT message, std::shared_ptr<IObserver<MT>> observer) {
+template <typename MT, typename ST>
+requires(Concepts::Observable<MT, ST>)
```
Here may be we can discuss where these functions are to be defined to avoid repeating this `concept` check, or perhaps there is a better solution?

tarun-mitruka commented on ikarus/controlroutines/loadcontrol.inl:
```diff
@@ -11,35 +11,33 @@
 
+  /// Dummy execution of the code with loadParameter as 0 inorder to save information of the undeformed (or initial)
+  /// configuration while using, for example, the class ControlSubsamplingVertexVTKWriter
   loadParameter = 0.0;
```
remove this

tarun-mitruka commented on ikarus/utils/observer/nonlinearsolverlogger.cpp:
```diff
@@ -8,48 +8,37 @@
-
-void NonLinearSolverLogger::updateImpl(NonLinearSolverMessages message, double val) {
-  switch (message) {
     case NonLinearSolverMessages::RESIDUALNORM_UPDATED:
```
Do you think it is nicer to remove these cases like `RESIDUALNORM_UPDATED` and similar which is doing nothing here in this file? The `default` behavior is similar to these cases anyways

rath3t commented on docs/website/01_framework/observer.md:
````diff
@@ -18,46 +18,64 @@ enum class NonLinearSolverMessages {
 ```
 
-The observer has to implement the function ``void updateImpl(MessageType message)``. In this function, all actions can
+The observer has to implement the function ``void updateImpl(MessageType message, const StateType& state)``.
````
One backtick is enough here.

rath3t commented on docs/website/01_framework/observer.md:
```diff
@@ -18,46 +18,64 @@ enum class NonLinearSolverMessages {
 be implemented that should be performed when the corresponding message is received.
 
-To connect observer and observable, one has to call ``observalbe.subscribe(MessageType::Message,observer)``. Example:
+To connect the observer and the observable, one has to call ``observalbe.subscribe(MessageType::Message, observer)``.
```
also here

rath3t replied:
also typo there. Actually al these descriptions should be removed here and to the doxygen documentation.
Basically you only have to move this text to the code, and add the doxygen commands, where the observer and observables are and remove the site here.

rath3t replied:
Also when describing this should be done without the observer object.

tarun-mitruka replied:
I fixed the backtick and the typo. Regarding moving these comments to the doxygen documentation, I would propose to do that in a separate PR, where the complete documentation is refactored such that only the framework is discussed in the documentation while such implementation aspects can be moved to doxygen documentation. For example, I think [here](https://ikarus-project.github.io/dev/01_framework/controlRoutines/#path-following-techniques) `SubsidiaryArgs` need not be documented as such because it is better mentioned [here](https://ikarus-project.github.io/dev/doxygen/html/a01221.html). What do you think?

tarun-mitruka replied:
I removed such lines

rath3t commented on docs/website/02_examples/vonMisesTruss.md:
```diff
@@ -59,12 +59,12 @@ This is implemented as depicted in the following:
-      lambdaAndDisp(1, step) = d[2];
-      lambdaAndDisp(2, step) = d[3];
-    });
+auto lvkObserver = std::make_shared<GenericObserver<ControlMessages, ControlState>>(
```
The observer should be able to get as a template just the observable object.

rath3t replied:
But I remember that we had reasons to keep the templates of the observer like this, I dont remember. Can you?

tarun-mitruka replied:
Ah yes you are right. I fixed this here `ikarus/utils/observer/genericobserver.hh` where we have 
```cpp
template <typename MT, typename ST>
class GenericObserver : public IObserver<IObservable<MT, ST>>
```
So accordingly in the example, observable must be passed. I will fix it

rath3t commented on ikarus/controlroutines/controlstate.hh:
```diff
@@ -0,0 +1,42 @@
+struct ControlState
+{
+  bool success{false};      ///< Flag indicating the success of the control.
+  bool initialConfig{true}; ///< Flag indicating if the state of deformation is in the initial configuration
```
To have this flag here, does not make sense.

tarun-mitruka replied:
I included this such that a user can check and log information accordingly [(example)](https://github.com/ikarus-project/ikarus/blob/81614f52be2e03f3663d579826dd68789679bf11/ikarus/utils/observer/controllogger.cpp#L31) and also perhaps it tells if the state of the control routine is in the initial configuration or not. 

May be instead of calling [`updateAndNotifyControlState(controlState, nonOp, solverState);`](https://github.com/ikarus-project/ikarus/blob/81614f52be2e03f3663d579826dd68789679bf11/ikarus/controlroutines/pathfollowing.inl#L46) here, once can just update the control state with information from the non-linear operator and just notify `SOLUTION_CHANGED` and remove `STEP_ENDED` (as you once suggested). Then this would get rid of the requirement of having this `bool initialConfig`.

rath3t replied:
What the bool is used for I saw in the code, but this is not a good design. There has to be a better way, [here](https://github.com/ikarus-project/ikarus/blob/81614f52be2e03f3663d579826dd68789679bf11/ikarus/utils/observer/controlvtkwriter.hh#L84-L95) the vtk writer can also write directly when CONTROL_STARTED is called or we may use the current step 0 as the initial step.

tarun-mitruka replied:
So something like this?
For example, in `ikarus/controlroutines/pathfollowing.inl` we have
```cpp
ControlState controlState{.success = false, .currentStep = 0, .stepSize = stepSize_, .name = this->name()};
  auto& nonOp = nonLinearSolver_->nonLinearOperator();
 controlState.sol    = &nonOp.firstParameter();
 controlState.lambda = nonOp.lastParameter();
 this->notify(ControlMessages::CONTROL_STARTED, controlState);
```

and then in `ikarus/utils/observer/controlvtkwriter.hh` we just have 
```cpp
case ControlMessages::CONTROL_STARTED:
case ControlMessages::SOLUTION_CHANGED: { ... } break;
```

rath3t commented on ikarus/controlroutines/controlstate.hh:
```diff
@@ -0,0 +1,42 @@
+  bool success{false};      ///< Flag indicating the success of the control.
+  bool initialConfig{true}; ///< Flag indicating if the state of deformation is in the initial configuration
+  int currentStep{-1};      ///< Information about the current step number.
+  int totalIterations{-1};  ///< Information about the total iterations performed.
```
It is not clear what iterations are meant. Do you mean the accumulated iterations of the non-linear solver?

rath3t replied:
If yes this information is not needed here. It can be collected from the `std::vector<NonLinearSolverState>`

tarun-mitruka replied:
Yes the accumulated iterations are meant here. This was existing in some previous [state](https://github.com/ikarus-project/ikarus/blob/bc987df2c0f29c29f8862459ecf0a72176c68dd5/ikarus/controlroutines/controlinfos.hh#L25), but you are right, it can be removed in this PR

rath3t commented on ikarus/controlroutines/controlstate.hh:
```diff
@@ -0,0 +1,42 @@
+  double stepSize{std::numeric_limits<double>::infinity()}; ///< Information about the step size.
+  double lambda{std::numeric_limits<double>::infinity()};   ///< Value of the load factor.
+  std::vector<NonLinearSolverState>
+      solverState{};                 ///< Vector containing information about the state of the nonlinear solver.
```
the variable name should be plural.

rath3t commented on ikarus/controlroutines/loadcontrol.inl:
```diff
@@ -11,35 +11,33 @@
-    return info;
-  this->notify(ControlMessages::SOLUTION_CHANGED);
-  this->notify(ControlMessages::STEP_ENDED);
+  /// Dummy execution of the code with loadParameter as 0 in order to save information of the undeformed (or initial)
```
can you remove the reference to undeformed. This is generic code. No need to write mechanic specific things.

rath3t commented on ikarus/solver/nonlinearsolver/newtonraphsonwithscalarsubsidiaryfunction.hh:
```diff
@@ -93,15 +94,17 @@ public:
-    solverInformation.success = true;
-    auto& x                   = nonLinearOperator().firstParameter(); // x = D (Displacements)
+    solverState.success = true;
+    auto& x             = nonLinearOperator().firstParameter(); // x = D (Displacements)
```
Wrong comment added here. (or too specific)

rath3t replied:
It was wrong before already. Can you remove it?

rath3t commented on ikarus/solver/nonlinearsolver/newtonraphsonwithscalarsubsidiaryfunction.hh:
```diff
@@ -169,23 +172,25 @@ public:
-      this->notify(NonLinearSolverMessages::CORRECTIONNORM_UPDATED, static_cast<double>(dNorm));
-      this->notify(NonLinearSolverMessages::RESIDUALNORM_UPDATED, static_cast<double>(rNorm));
-      this->notify(NonLinearSolverMessages::ITERATION_ENDED);
+      solverState.residualNorm   = static_cast<double>(rNorm);
```
Why casting from double to double?

tarun-mitruka replied:
If I remember correctly, `auto rNorm = sqrt(rx.dot(rx));` is not always of `double` type because when automatic differentiation is used, it has a different type there. Hence I think this casting step was necessary also for [`NewtonRaphson`](https://github.com/ikarus-project/ikarus/blob/bc987df2c0f29c29f8862459ecf0a72176c68dd5/ikarus/solver/nonlinearsolver/newtonraphson.hh#L129)

rath3t replied:
Ah yes you are right.

rath3t commented on ikarus/solver/nonlinearsolver/solverstate.hh:
```diff
@@ -0,0 +1,39 @@
+  double correctionNorm{std::numeric_limits<double>::infinity()}; ///< Value of the correction norm.
+
+  // optional variables specific for the solver NewtonRaphsonWithSubsidiaryFunction
+  std::optional<double> lambda;             ///< Value of the load factor.
```
Having optionals for different solver states does not make sense, since it will not be scalable. Maybe you can think of a better solution to print the values? Maybe with dynamic polymorphism or if the state is a template, w.r.t. the nonlinear solver. Adding fields here, is not useful.

rath3t commented on ikarus/solver/nonlinearsolver/trustregion.hh:
```diff
@@ -332,29 +332,31 @@ public:
+        solverState.residualNorm   = stats_.gradNorm;
+        this->notify(NonLinearSolverMessages::CORRECTIONNORM_UPDATED, solverState);
+        this->notify(NonLinearSolverMessages::RESIDUALNORM_UPDATED, solverState);
+        this->notify(NonLinearSolverMessages::SOLUTION_CHANGED, solverState);
```
Did you check that the output stays as before?

tarun-mitruka replied:
I executed `tests/src/testnonlinearelasticitysvk.cpp` now in both `main` and this branch and have attached their output below:

[main.txt](https://github.com/ikarus-project/ikarus/files/14840820/main.txt)
[refactor_control_loggers.txt](https://github.com/ikarus-project/ikarus/files/14840822/refactor_control_loggers.txt)

The output looks the same. The only difference is the extra unnecessary number of lines printed for the result from the `main` branch because there non-linear solver is subscribed for the undeformed state which is fixed in this PR

tarun-mitruka replied:
Furthermore, the logging information itself isn't modified here because the trust region method has its own logging information and its implementation will be synchronized in a separate PR

rath3t commented on ikarus/utils/concepts.hh:
```diff
@@ -198,6 +201,19 @@ namespace Concepts {
+  template <typename MT, typename ST>
+  concept Observable = requires {
+    (std::is_same_v<MT, ControlMessages> && std::is_same_v<ST, ControlState>) ||
+        (std::is_same_v<MT, NonLinearSolverMessages> && std::is_same_v<ST, NonLinearSolverState>);
```
This concept does not make sense, since it is not scalable. You should check, if `MT` and `ST` adhere another concept and not check that they are of a specific type.

tarun-mitruka replied:
Did you mean like this? https://godbolt.org/z/n9r3Ebrfn

rath3t commented on tests/src/testadaptivestepsizing.cpp:
```diff
@@ -252,12 +253,12 @@ int main(int argc, char** argv) {
   const std::vector<std::vector<int>> expectedIterationsALC = {
-      {9, 9, 6, 5, 4, 4},
-      {9, 8, 6, 5, 5, 5}
+      {0, 9, 9, 6, 5, 4, 4},
```
Ok is this what we expect? Shouldn't this be like the old? Im not sure what would be better.

tarun-mitruka replied:
In some intermediate state of this PR, for the undeformed configuration, the nonlinear solver was still used to solve the complete system of equations which led to zero iterations, but you are right, the current version removes this step and hence this should go back to the old state. This also leads to some changes in the control routine which I will simplify

rath3t reviewed (commented):
There are several minor things to change but overall I think we have to refactor the observers to have nice solution, which is more scalable. First, we have to decide, which state contains what information and then we can refactor from there.

Currently several workaround are in the code I would consider code smells, which indicate there is a nicer solution.

tarun-mitruka commented on ikarus/controlroutines/loadcontrol.inl:
```diff
@@ -11,35 +11,28 @@
+  this->notify(ControlMessages::CONTROL_STARTED, controlState);
   auto& loadParameter = nonOp.lastParameter();
 
-  loadParameter = 0.0;
-  this->notify(ControlMessages::STEP_STARTED, 0, stepSize_);
-  auto solverInfo = nonLinearSolver_->solve();
```
Since we removed the solve step here, do you think it is nice to check if the norm of the residuum is zero at this state, before any increment happens? Or is it a hard constraint that the input state to a control routine should be in equilibrium?

tarun-mitruka committed:
refactor control and nonlinear solver logger and controlvtkwriters

refactor observer, observable, genericobserver

refactor solverState

refactor control state

add a concept for observable

fix and refactor tests

update docs

update CHANGELOG

minor fixes

tarun-mitruka committed:
Update trustregion.hh
